framework/src/problems/FEProblemBase.C

```diff
@@ -2225,8 +2225,13 @@ FEProblemBase::addInitialCondition(const std::string & ic_name,
       MooseVariableFEBase & var = getVariable(
           tid, var_name, Moose::VarKindType::VAR_ANY, Moose::VarFieldType::VAR_FIELD_ANY);
       parameters.set<SystemBase *>("_sys") = &var.sys();
-      std::shared_ptr<InitialCondition> ic =
-          _factory.create<InitialCondition>(ic_name, name, parameters, tid);
+      std::shared_ptr<InitialConditionBase> ic;
+      if (dynamic_cast<MooseVariable *>(&var))
+        ic = _factory.create<InitialCondition>(ic_name, name, parameters, tid);
+      else if (dynamic_cast<VectorMooseVariable *>(&var))
+        ic = _factory.create<VectorInitialCondition>(ic_name, name, parameters, tid);
+      else
+        mooseError("Your FE variable must be either of scalar or vector type");
```

Comment:
I am not sure if I missed this first time, but this needs to be improved. At least use `name` in this error message, so that people know which block in their input file is throwing this error. Otherwise, they will have no idea what is causing this error and how to fix it.